src/fu-engine-repair.c

```diff
@@ -0,0 +1,441 @@
+
+	grubby = fu_path_find_program("grubby", NULL);
+	argv_grubby[0] = grubby;
+
+	if (enable)
+		arg_string = g_strdup_printf("--args=%s", grubby_arg);
+	else
+		arg_string = g_strdup_printf("--remove-args=%s", grubby_arg);
+
+	argv_grubby[2] = arg_string;
+
+	if (!g_spawn_sync(NULL,
+			  (gchar **)argv_grubby,
+			  NULL,
+			  G_SPAWN_DEFAULT,
+			  NULL,
+			  NULL,
+			  &output,
+			  NULL,
+			  NULL,
+			  error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+is_kernel_parameter(const gchar *para_string, GError **error)
+{
+	g_autofree gchar *kernel_cmdline = NULL;
+	gsize length;
+
+	if (!g_file_get_contents("/proc/cmdline", &kernel_cmdline, &length, error)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR_INTERNAL,
+				    FWUPD_ERROR_READ,
+				    "Fail on reading kernel parameter.");
+		return FALSE;
+	}
+
+	if (g_str_match_string(para_string, kernel_cmdline, TRUE))
+		return TRUE;
+
+	return FALSE;
+}
+
+static gboolean
+grubby_set_lockdown(gboolean enable, GError **error)
+{
+	if (enable) {
+		return grubby_set(TRUE, "lockdown=confidentiality", error);
+	} else {
+		if (!is_kernel_parameter("lockdown=", error)) {
+			g_set_error_literal(
+			    error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "Can't be reverted since kernel lockdown was disabled.");
+			return FALSE;
+		}
+
+		return grubby_set(FALSE, "lockdown=confidentiality", error);
+	}
+}
+
+static gboolean
+grubby_set_iommu(gboolean enable, GError **error)
+{
+	gboolean ret;
+	if (enable) {
+		return grubby_set(TRUE, "iommu=force", error);
+	} else {
+		ret = is_kernel_parameter("iommu=force", error);
+		if (!*error && ret) {
+			g_printf("OK iommu ret %d\n", ret);
+			return grubby_set(FALSE, "iommu=force", error);
+		} else {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_NOTHING_TO_DO,
+					    "iommu was not set.");
+			return FALSE;
+		}
+	}
+}
+
+static gboolean
+fu_engine_repair_kernel_lockdown(FuEngine *engine, const gchar *action, GError **error)
+{
+	FuSecurityAttrs *attrs;
+	FwupdSecurityAttr *attr;
+	guint flags;
+
+	attrs = fu_engine_get_host_security_attrs(engine);
+	if (!attrs)
+		return FALSE;
+
+	attr = fu_security_attrs_get_by_appstream_id(attrs, FWUPD_SECURITY_ATTR_ID_UEFI_SECUREBOOT);
+	if (!attr)
+		return FALSE;
+
+	flags = fwupd_security_attr_get_flags(attr);
+
+	if (!g_strcmp0(action, "undo")) {
+		if (flags == FWUPD_SECURITY_ATTR_FLAG_SUCCESS) {
+			g_set_error_literal(
+			    error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOTHING_TO_DO,
+			    "Kernel lockdown can't be disabled when secure boot is enabled.");
+			return FALSE;
+		} else {
+			return grubby_set_lockdown(FALSE, error);
+		}
+	}
+
+	return grubby_set_lockdown(TRUE, error);
+}
+
+static gboolean
+fu_engine_repair_iommu(const gchar *action, GError **error)
+{
+	if (!is_grubby_installed(error)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    FWUPD_ERROR_NOTHING_TO_DO,
```

Comment:
I don't think this should be nothing to do. That's going to return error code 2 when this is actually fatal.

Reply:
Thank you for addressing it. I'll pick up an appropriate error code for them.